lib/aliquot/payment.rb

```diff
@@ -84,47 +82,153 @@ def protocol_version
+      validator.validate
+
+      validator.output
+    end
+
+    def intermediate_key_expired?
+      cur_millis = (Time.now.to_f * 1000).round
+      @intermediate_key[:keyExpiration].to_i < cur_millis
     end
 
-    def valid_signature?
-      signed_string = ['Google', @merchant_id, protocol_version, @token[:signedMessage]].map do |str|
+    def check_shared_secret
+      begin
+        decoded = Base64.strict_decode64(@shared_secret)
+      rescue
+        raise InvalidSharedSecretError, 'shared_secret must be base64'
+      end
+
+      raise InvalidSharedSecretError, 'shared_secret must be 32 bytes when base64 decoded' unless decoded.length == 32
+    end
+
+    def valiate_signature
+      signed_string_message = ['Google', @merchant_id, protocol_version, @token[:signedMessage]].map do |str|
         [str.length].pack('V') + str
       end.join
+      message_signature = Base64.strict_decode64(@token[:signature])
 
-      keys = JSON.parse(@signing_keys)['keys']
-      # Check if signature was performed with any possible key.
-      keys.map do |key|
-        next if key['protocolVersion'] != protocol_version
+      root_signing_keys = JSON.parse(@signing_keys)['keys'].select do |key|
+        key['protocolVersion'] == protocol_version
+      end
+
+      root_signing_keys.map! do |key|
+        OpenSSL::PKey::EC.new(Base64.strict_decode64(key['keyValue']))
+      end
+
+      case protocol_version
+      when 'ECv1'
+        # Check if signature was performed directly with any possible key.
+        success =
+          root_signing_keys.map do |key|
+            key.verify(digest, message_signature, signed_string_message)
+          end.any?
+
+        raise InvalidSignatureError unless success
+      when 'ECv2'
+        signed_key_signature = ['Google', 'ECv2', @token[:intermediateSigningKey][:signedKey]].map do |str|
+          [str.length].pack('V') + str
+        end.join
+
+        # Check that the intermediate key signed the message
+        pkey = OpenSSL::PKey::EC.new(Base64.strict_decode64(@intermediate_key[:keyValue]))
+        raise InvalidSignatureError, 'intermediate did not sign message' unless pkey.verify(digest, message_signature, signed_string_message)
+
+        intermediate_signatures = @token[:intermediateSigningKey][:signatures]
+
+        success = valid_intermediate_key_signatures?(
+          root_signing_keys,
+          intermediate_signatures,
+          signed_key_signature
+        )
+
+        raise InvalidSignatureError, 'intermediate not signed' unless success
+      end
+    end
 
-        ec = OpenSSL::PKey::EC.new(Base64.strict_decode64(key['keyValue']))
-        ec.verify(OpenSSL::Digest::SHA256.new, Base64.strict_decode64(@token[:signature]), signed_string)
+    def valid_intermediate_key_signatures?(signing_keys, signatures, signed)
+      signing_keys.map do |key|
+        signatures.map do |sig|
+          key.verify(digest, Base64.strict_decode64(sig), signed)
```

Comment:
Yeah, it's sneaky.

It's the same with `cipher`, where it's slightly less magical since it actually depends on the `protocolVersion`.
I considered a function such as `get_cipher`, but it breaks the style rule Naming/AccessorMethodName, arbitrary reason, i know.

I'm actively looking for a better way to have these class wide, so I'm up for ideas :)

Reply:
(:heavy_check_mark:)